python/raft/common/handle.pyx

```diff
@@ -50,51 +51,36 @@ cdef class Handle:
         del handle  # optional!
     """
 
-    # handle_t doesn't have copy operator. So, use pointer for the object
-    # python world cannot access to this raw object directly, hence use
-    # 'size_t'!
-    cdef size_t h
-
-    # not using __dict__ unless we need it to keep this Extension as lean as
-    # possible
-    cdef int n_streams
-
-    def __cinit__(self, n_streams=0):
+    def __cinit__(self, stream=None, n_streams=0):
```

Comment:
Eventually it would also be nice to either have the Python handle accept other library resources explicitly or to interact directly w/ or wrap the cupy-maintained resources. Currently, we just use the default stream in cupy because there's no sharing/interaction between the resources held by cupy and cuml.

Reply:
Interested how / if cugraph is also handling this (cc @Iroy30)

Reply:
Yeah, cupy has a stream setter as well, which acts like a singleton I believe. You can set it with the pointer of the stream that you own. We could set it with the stream used by the handle, I'm not sure how to correctly context manage it though. Maybe it should be context managed by cuML or cuGraph when they get the handle and want to run cupy stuff.

https://docs.cupy.dev/en/stable/reference/generated/cupy.cuda.ExternalStream.html